### PR TITLE
Fix the way to get `state` in beacon block topic validator

### PR DESCRIPTION
**Pull request by NIC619**
### What was wrong?
L45: `state = chain.get_state_by_slot(block.slot - 1)` will raise error if the state of that slot is not persisted or if the slot is in the future.

### How was it fixed?
We need to get the state with `state.slot < block.slot` where `block` is the block to be validated. So we first check if head state slot is less than `block.slot`, if not then we check head block slot.
If both slots are greater than `block.slot` then we check the ancestor blocks of head block.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [ ] Clean up commit history

[//]: # (For important changes that should go into the release notes please add a newsfragment file as explained here: https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

[//]: # (See: https://trinity-client.readthedocs.io/en/latest/contributing.html#pull-requests)
- [ ] Add entry to the [release notes](https://github.com/ethereum/trinity/blob/master/newsfragments/README.md)

### CODE REVIEW

File: trinity/protocol/bcc_libp2p/topic_validators.py [hunk 1]
```diff
@@ -29,6 +35,59 @@
...
+    latest_finalized_slot = compute_start_slot_of_epoch(
+        state.finalized_checkpoint.epoch,
+        state_machine.config.SLOTS_PER_EPOCH,
+    )
+    if block.slot < latest_finalized_slot:
+        raise ValidationError("block is older than latest finalized slot")
+
+    # Find the proper state to verify the proposer's signature of the block
+    # head state slot >= target block slot, try head block
+    if state.slot >= block.slot:
```
Comment (ralexstokes) [inline trinity/protocol/bcc_libp2p/topic_validators.py]: why would we not just get the state at the `block`'s `slot`? i don't see why we need to pull any beacon blocks out of our local state here... i'll move my thinking to the top-level comment on this review

**Comment by NIC619** [comment]
> 1. keeping the `process_block_header` interface

One problem is that the `validate_block_parent_root` check in `process_block_header`:
https://github.com/ethereum/trinity/blob/96cb38c41a8d22b8d0d7f18132b8bbe0f38cd085/eth2/beacon/state_machines/forks/serenity/block_validation.py#L80-L87

This would require us to pass in the correct state, i.e., pass in the fork chain state when validating a fork block, so we need to fetch state by `block.state_root` instead of by `slot`. (I realize this part: `if state.slot >= block.slot:` is wrong in this case)

> why would we not just get the state at the `block`'s `slot`?

See the reason above and also the stale/incorrect state problem with `get_state_by_slot` mentioned in #1165 .

> there is another concern here that we may have a synchronization issue b/t the Chain and this code, i.e. the various calls we make while interacting w/ the internals of the beacon chain could change out from under us, invalidating the remainder of this computation. 

Good point! This is definitely a problem and need to be handled carefully...

**Comment by NIC619** [comment]
This problem is way trickier than I thought. @ralexstokes @hwwhww @ChihChengLiang What do you think about changing it to 
1. first get the state of the block's parent block, apply state transition and use it to verify the block
2. if no parent block is found, skip validation(IIUC this is how lighthouse handles it **_for now_**)

**Comment by NIC619** [comment]
Fixed by #1427

**Pull request closed without merging**